### PR TITLE
CARBON-886 Reorganise dependencies

### PR DESCRIPTION
**Pull request by ianoxley**
Add the following to `peerDependencies` and set their version to `~15.6.1` (the same as the version specified for these packages in `devDependencies`):

* react
* react-dom
* flux

**NB** `react`, `react-dom`, and `flux` are still specified in `devDependencies`.

Move the following from `devDependencies` to `dependencies`:

* i18n-js
* immutable
* react-router
* react-transition-group

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -21,26 +21,24 @@
...
+    "react-dom": ">= 15.5.4 < 16"
+  },
   "devDependencies": {
     "babel-standalone": "~6.17.0",
     "carbon-factory": "1.1.7",
     "enzyme": "~2.8.2",
     "express": "~4.14.0",
-    "flux": "~2.1.1",
     "gulp": "~3.9.0",
     "highcharts": "~5.0.2",
```
Comment (SeanArmstrong) [inline package.json]: Hightcharts will also need moving?

**Comment by ianoxley** [comment]
`highcharts` only appears to be used in the demo site, and is currently in `devDependencies`. It might be worth removing it, as `react-highcharts` is in `dependencies`.

`form-serialize` is already in `dependencies`.

**Comment by ianoxley** [comment]
It turns out there's some spec helper that uses `highcharts` so I'm leaving it in `devDependencies`.

File: package.json [hunk 2]
```diff
@@ -21,6 +21,11 @@
     "url": "https://github.com/Sage/carbon/issues"
   },
   "homepage": "https://github.com/Sage/carbon#readme",
+  "peerDependencies": {
+    "flux": "~2.1.1",
+    "react": "~15.6.1",
```
Comment (andrewjtait) [inline package.json]: Could these be dependencies? While it would remove any flexibility for consumers to use different versions of React, we wouldn't really support any other version of React (eg they couldn't use v16 if we didn't yet support it).

It would also mean that if we control the version we could ensure that applications remain up to date with minor versions so consumers don't have to worry about it so much.

On the other side, if we controlled the version we're not guaranteeing that consumers will remain up to date with Carbon - so if they are using a version of Carbon locked to version 15.6.0 and they need to use 15.7.0, they may need to do a lot of upgrading of Carbon versions (which may or may not be a bad thing).

Peer dependency is useful for modules which are supplying a small piece to an application, whereas Carbon is more of a library for building and managing your application which is why I don't see declaring them as dependencies as such a negative issue.

Interested to hear opinions on what people would prefer.

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,20 @@
...
+* react
+* react-dom
+* flux
+
+The following packages have been moved from `devDependencies` to `dependencies`:
+
+* `i18n-js`
+* `immutable`
+* `react-router`
+* `react-transition-group`
```
Comment (andrewjtait) [inline CHANGELOG.md]: It would be useful to have guides on how to upgrade to this version (eg. what do they need to change in their `package.json` file?)

File: package.json [hunk 4]
```diff
@@ -64,6 +67,8 @@
     "react-dnd-touch-backend": "lock this to 0.2.7 to support dnd on mobile - see https://github.com/yahoo/react-dnd-touch-backend/issues/34",
     "react-dnd-touch-backend": "0.2.7",
     "react-highcharts": "~11.5.0",
+    "react-router": "~2.8.0",
```
Comment (andrewjtait) [inline package.json]: Interestingly GAC uses v3 - should we upgrade it to that?

File: package.json [hunk 5]
```diff
@@ -55,6 +56,8 @@
     "escape-string-regexp": "~1.0.5",
     "events": "~1.1.1",
     "form-serialize": "~0.7.0",
+    "i18n-js": "http://github.com/fnando/i18n-js/archive/v3.0.0.rc8.tar.gz",
```
Comment (andrewjtait) [inline package.json]: please use RC12

File: package.json (hunk 4 above)
Comment (SeanArmstrong) [inline package.json]: Yeah there shouldn't be too much fallout in 3 and it should catch all of the proptype warnings. v4 on the other hand is a big jump and refactor

Comment (SeanArmstrong) [inline package.json]: This to be done as a separate PR?

**Comment by ianoxley** [comment]
Updated Flux to `^3.0.0`

Moved `prop-types` to `dependencies` (Facebook recommends adding `prop-types` as a dependency for libraries).

**Comment by toni-leigh** [comment]
Could have raised this as an issue and kept the work fully visible rather than obfuscated behind a ticket number - ticket then would link to the issue for internal tracking

File: CHANGELOG.md (hunk 3 above)
Comment (ianoxley) [inline CHANGELOG.md]: Is this what you had in mind
https://github.com/Sage/carbon/blob/efe7142e41c37cbdca0222179f1441a1b08ce7cc/CHANGELOG.md#upgrading-a-project-that-uses-carbon

**Comment by ianoxley** [comment]
`react-router` now on `^3.0.0`

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,34 @@
...
+
+The following packages have been upgraded:
+
+* flux: ^3.0.0
+* i18n-js: upgraded to rc12
+* react-router: ^3.0.0
+
+### Upgrading a project that uses Carbon
+
+If you're upgrading an application that uses Carbon to v2.0.0 you'll need to make sure you have `flux`, `react`, and `react-dom` in your project's dependencies. If you need to add them you can do so by running the following command:
```
Comment (andrewjtait) [inline CHANGELOG.md]: If they're already using Carbon they'll already have these dependencies installed, we should tell them how they need to change their package.json

* they should delete the modules which are now dependencies.
* they should upgrade the version of flux to `^3.1.1`.

File: package.json [hunk 7]
```diff
@@ -21,26 +21,27 @@
...
+    "react": "^15.6.1",
+    "react-dom": "^15.6.1"
+  },
   "devDependencies": {
     "babel-standalone": "~6.17.0",
     "carbon-factory": "1.1.7",
     "enzyme": "~2.8.2",
     "express": "~4.14.0",
-    "flux": "~2.1.1",
+    "flux": "^3.0.0",
```
Comment (andrewjtait) [inline package.json]: why don't we use 3.1.1?

File: package.json [hunk 8]
```diff
@@ -21,26 +21,27 @@
     "url": "https://github.com/Sage/carbon/issues"
   },
   "homepage": "https://github.com/Sage/carbon#readme",
+  "peerDependencies": {
+    "flux": "^3.0.0",
```
Comment (andrewjtait) [inline package.json]: why don't we use 3.1.1?

File: src/components/i18n/__spec__.js [hunk 9]
```diff
@@ -95,7 +95,13 @@ describe('I18n', () => {
   });
 
   describe("tags on component", () => {
-    let wrapper = shallow(<I18n data-element='bar' data-role='baz' />);
+    let wrapper = shallow(
+      <I18n 
+        scope='foo'
```
Comment (andrewjtait) [inline src/components/i18n/__spec__.js]: has this been added accidently?

File: package.json (hunk 8 above)
Comment (ianoxley) [inline package.json]: We kind of do, as running `npm install` will install the latest v3 Flux i.e. 3.1.2.

However, is it better to offer more flexibility and say Carbon will work with any v3 of Flux? Then other apps don't have to update Flux as often.

File: package.json (hunk 7 above)
Comment (ianoxley) [inline package.json]: See above comment on peer dependencies (this is set to match whatever we go with in peer dependencies).

File: package.json (hunk 8 above)
Comment (andrewjtait) [inline package.json]: OK - I just wanted to ensure that applications are encouraged to upgrade. I don't expect them to be setting `^3.0.0` but instead to be locking their versions so they don't get unexpected upgrades. Since `3.1.1` had no meaningful updates it didn't seem like a bad idea to set that as the minimum.

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,34 @@
...
+
+```
+npm install flux@^3.0.0 react react-dom --save
```
Comment (andrewjtait) [inline CHANGELOG.md]: they should install to specific versions, to avoid unintended upgrades (ie they could get a new version during a deployment which could be bad)

**Comment by andrewjtait** [comment]
added the do not merge label as it is targeted for v2.

**Comment by andrewjtait** [comment]
Tested in demo site and web app successfully ✅
